compiler-plugin/src/main/kotlin/arrow/meta/quotes/parentscopes/ClassOrObjectScope.kt

```diff
@@ -0,0 +1,51 @@
+package arrow.meta.quotes.parentscopes
+
+import arrow.meta.quotes.ClassBodyScope
+import arrow.meta.quotes.Scope
+import arrow.meta.quotes.ScopedList
+import com.intellij.navigation.ItemPresentation
+import org.jetbrains.kotlin.name.Name
+import org.jetbrains.kotlin.psi.KtAnnotationEntry
+import org.jetbrains.kotlin.psi.KtAnonymousInitializer
+import org.jetbrains.kotlin.psi.KtClass
+import org.jetbrains.kotlin.psi.KtClassBody
+import org.jetbrains.kotlin.psi.KtClassOrObject
+import org.jetbrains.kotlin.psi.KtObjectDeclaration
+import org.jetbrains.kotlin.psi.KtDeclaration
+import org.jetbrains.kotlin.psi.KtModifierList
+import org.jetbrains.kotlin.psi.KtParameter
+import org.jetbrains.kotlin.psi.KtParameterList
+import org.jetbrains.kotlin.psi.KtPrimaryConstructor
+import org.jetbrains.kotlin.psi.KtSuperTypeList
+import org.jetbrains.kotlin.psi.KtSuperTypeListEntry
+import org.jetbrains.kotlin.psi.allConstructors
+import org.jetbrains.kotlin.psi.getOrCreateBody
+import org.jetbrains.kotlin.psi.psiUtil.modalityModifierType
+
+/**
+ * A template destructuring [Scope] for a [KtClassOrObject]
+ *
+ * Parent scope of [KtClass] and [KtObjectDeclaration]
+ */
+open class ClassOrObjectScope<out T : KtClassOrObject>(
+  override val value: T,
+  val `@annotationEntry`: ScopedList<KtAnnotationEntry> = ScopedList(value.annotationEntries),
+  val modality: Name? = value.modalityModifierType()?.value?.let(Name::identifier),
+  val superTypeList: KtSuperTypeList? = value.getSuperTypeList(),
+  val superTypeListEntries: ScopedList<KtSuperTypeListEntry> = ScopedList(prefix = "<", value = value.superTypeListEntries, postfix = ">"),
+  val anonymousInitializers: ScopedList<KtAnonymousInitializer> = ScopedList(value = value.getAnonymousInitializers(), postfix = ","),
+  val body: ClassBodyScope = ClassBodyScope(value.body),
+  val declarations: ScopedList<KtDeclaration> = ScopedList(value = value.declarations, postfix = ", "),
+  val presentation: ItemPresentation = value.presentation,
+  val primaryConstructor: KtPrimaryConstructor? = value.primaryConstructor,
+  val primaryConstructorModifierList: KtModifierList? = primaryConstructor?.modifierList,
+  val primaryConstructorParameterList: KtParameterList? = primaryConstructor?.valueParameterList,
```

Comment:
@raulraja @bloderxd What should we be doing with KtElementList types?

The parameter list doesn't seem to matter as much, but `KtModifierList` has methods like:
```
    @NotNull
    public List<KtAnnotation> getAnnotations() {
        return getStubOrPsiChildrenAsList(KtStubElementTypes.ANNOTATION);
    }

    @NotNull
    public List<KtAnnotationEntry> getAnnotationEntries() {
        return KtPsiUtilKt.collectAnnotationEntriesFromStubOrPsi(this);
    }
```

Reply:
@bloderxd and myself explored all this and we found that much of this is unnecessary boxing. Changing it to all use `@annotation` for the type `KtAnnotationEntry` and for all lists of KtAnnotation, `ScopedList<KtAnnotationEntry>`